src/SampleAggregator.cpp

```diff
@@ -61,108 +58,320 @@ namespace geopm
+                m_platform_io.push_signal("EPOCH_COUNT", domain_type, domain_idx),
+                0,
+                AvgAccumulator::make_unique(),
+                AvgAccumulator::make_unique(),
+                {},
+                {},
+           };
+        }
+        return result;
+    }
+
+
+    /// @todo Repeated code could be avoided with template
+    static std::map<uint64_t, std::shared_ptr<SumAccumulator> >::iterator
+    emplace_hash(
+        std::map<uint64_t, std::shared_ptr<SumAccumulator> > &hash_accum_map,
+        uint64_t hash)
+    {
+        std::map<uint64_t, std::shared_ptr<SumAccumulator> >::iterator result;
+        result = hash_accum_map.find(hash);
+        if (result == hash_accum_map.end()) {
+            auto empl_ret = hash_accum_map.emplace(hash, SumAccumulator::make_unique());
+            result = empl_ret.first;
+        }
+        return result;
+    }
+
+    static std::map<uint64_t, std::shared_ptr<AvgAccumulator> >::iterator
+    emplace_hash(
+        std::map<uint64_t, std::shared_ptr<AvgAccumulator> > &hash_accum_map,
+        uint64_t hash)
+    {
+        std::map<uint64_t, std::shared_ptr<AvgAccumulator> >::iterator result;
+        result = hash_accum_map.find(hash);
+        if (result == hash_accum_map.end()) {
+            auto empl_ret = hash_accum_map.emplace(hash, AvgAccumulator::make_unique());
+            result = empl_ret.first;
+        }
+        return result;
+    }
+
+    void SampleAggregatorImp::update_total(void)
+    {
+        // Update all of the sum aggregators
+        for (auto &signal_it : m_sum_signal) {
+            int signal_idx = signal_it.first;
+            m_sum_signal_s &signal = signal_it.second;
+            double sample = m_platform_io.sample(signal_idx);
+            uint64_t hash = m_platform_io.sample(signal.region_hash_idx);
+            int epoch_count = m_platform_io.sample(signal.epoch_count_idx);
+            if (!m_is_updated) {
+                // On first call just initialize the signal values
+                signal.sample_last = sample;
+                signal.region_hash_last = hash;
+                signal.epoch_count_last = epoch_count;
+                signal.region_accum_it = emplace_hash(signal.region_accum, hash);
+            }
+            else {
+                // Measure the change since the last update
+                double delta = sample - signal.sample_last;
+                // Update that application totals
+                signal.app_accum->update(delta);
+                // If the epoch count has changed, call the exit/enter
+                if (epoch_count != signal.epoch_count_last) {
+                    if (signal.epoch_count_last != 0) {
+                        signal.epoch_accum->exit();
+                    }
+                    signal.epoch_accum->enter();
+                }
+                // If we have observed our first epoch, update epoch totals
+                if (signal.epoch_count_last != 0) {
```

Comment:
doesn't this skip updating for the first sample in the first epoch, since epoch is not updated until 206?

Reply:
Same logic as for region entry/exit